codalab/model/bundle_model.py

```diff
@@ -830,19 +770,45 @@ def resume_bundle(self, bundle, bundle_update, row, user_id, worker_id, connecti
 
         return True
 
-    def finalize_bundle(self, bundle, user_id, exitcode, failure_message, connection):
+    def transition_bundle_worker_offline(self, bundle):
         """
-        Marks the bundle as READY / KILLED / FAILED, updating a few metadata fields and
-        removing the worker_run row. Additionally, if the user
-        running the bundle was the CodaLab root user, increments the time
-        used by the bundle owner.
+        Transitions bundle to WORKER_OFFLINE state:
+            Updates the last_updated metadata.
+            Removes the corresponding row from worker_run if it exists.
+        """
+        with self.engine.begin() as connection:
+            # Check that it still exists and is running
+            row = connection.execute(
+                cl_bundle.select().where(
+                    cl_bundle.c.id == bundle.id
+                    and (cl_bundle.c.state == State.RUNNING or cl_bundle.c.state == State.PREPARING)
+                )
+            ).fetchone()
+            if not row:
+                # The user deleted the bundle or the bundle finished
+                return False
+
+            # Delete row in worker_run
+            connection.execute(
+                cl_worker_run.delete().where(cl_worker_run.c.run_uuid == bundle.uuid)
+            )
+
+            bundle_update = {
+                'state': State.WORKER_OFFLINE,
+                'metadata': {'last_updated': int(time.time())},
+            }
+            self.update_bundle(bundle, bundle_update, connection)
+        return True
+
+    def transition_bundle_finalizing(self, bundle, user_id, exitcode, failure_message, connection):
+        """
+        Transitions bundle to FINALIZING state:
+            Saves the failure message and exit code from the worker
+            If the user running the bundle was the CodaLab root user,
+            increments the time used by the bundle owner.
         """
-        state = State.FAILED if failure_message or exitcode else State.READY
```

Comment:
This `state` var wasn't used in this function since this is for finalizing, the state is actually computed in the `finish_bundle` (now `transition_bundle_finished` function)